go-controller/pkg/crd/userdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,248 @@
+	// Allowed value is "Secondary".
+	// Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.
+	// +kubebuilder:validation:Required
+	// +required
+	Role NetworkRole `json:"role"`
+
+	// mtu is the maximum transmission unit for a network.
+	// MTU is optional, if not provided, the globally configured value in OVN-Kubernetes (defaults to 1400) is used for the network.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65536
+	// +optional
+	MTU int32 `json:"mtu,omitempty"`
+
+	// subnets are used for the pod network across the cluster.
+	// Dual-stack clusters may set 2 subnets (one for each IP family), otherwise only 1 subnet is allowed.
+	//
+	// The format should match standard CIDR notation (for example, "10.128.0.0/16").
+	// This field may be omitted. In that case the logical switch implementing the network only provides layer 2 communication,
+	// and users must configure IP addresses for the pods. As a consequence, Port security only prevents MAC spoofing.
+	// +kubebuilder:validation:MaxItems=2
+	// TODO: CIDR validation
+	// +optional
+	Subnets []string `json:"subnets,omitempty"`
+
+	// joinSubnets are used inside the OVN network topology.
+	// Dual-stack clusters may set 2 subnets (one for each IP family), otherwise only 1 subnet is allowed.
+	// This field is only allowed for "Primary" network.
+	// It is not recommended to set this field without explicit need and understanding of the OVN network topology.
+	// When omitted, the platform will choose a reasonable default which is subject to change over time.
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=2
+	// TODO: CIDR validation
+	// +optional
+	JoinSubnets []string `json:"joinSubnets,omitempty"`
+
+	// ipamLifecycle controls IP addresses management lifecycle.
+	// The only allowed value is Persistent. When set, OVN Kubernetes assigned IP addresses will be persisted in an
+	// `ipamclaims.k8s.cni.cncf.io` object. These IP addresses will be reused by other pods if requested.
+	// Only supported when "subnets" are set.
+	// +optional
+	IPAMLifecycle NetworkIPAMLifecycle `json:"ipamLifecycle,omitempty"`
+}
+
+// +kubebuilder:validation:XValidation:rule="self.role == 'Secondary'", message="LocalNet topology is only supported for Secondary network"
+// +kubebuilder:validation:XValidation:rule="!has(self.excludeSubnets) || has(self.subnets) && size(self.subnets) > 0", message="ExcludeSubnets is only supported when Subnets is provided"
+// +kubebuilder:validation:XValidation:rule="!has(self.ipamLifecycle) || has(self.subnets) && size(self.subnets) > 0", message="IPAMLifecycle is only supported when subnets are set"
+type LocalNetConfig struct {
+	// role describes the network role in the pod.
+	// Allowed values are "Primary" and "Secondary".
+	// Must be set to "Secondary".
+	// +kubebuilder:validation:Required
+	// +required
+	Role NetworkRole `json:"role"`
```

Comment:
if we only support this to be `secondary` ... 

Why even have it ?

Reply:
yeah that is something I wanted to discuss: Do we want to make it similar to other configs and explicitly request to set Secondary as network role, or just hide it and say always secondary

Reply:
I also think it may be easier in the future if we ever add new roles allowed for localnet to just add allowed values

Reply:
apparently we want to support this as primary in the future, so I think it validates the point ^

Reply:
Please note that if Role field is extracted and be part of spec we can avoid this corner.
In addition `spec` should have validation rule rejecting secondary with localnet topology spec, once primary:localnet is supported the rule can be dropped and such request is  accepted.

Reply:
If that's on the roadmap then moving it up to spec may make sense. Will there ever be more types of network? Would they always support primary/secondary semantics?

Reply:
> If that's on the roadmap then moving it up to spec may make sense
@trozet does primary network over localnet it planned to be supported in the future?

> Will there ever be more types of network? Would they always support primary/secondary semantics?
I think the interface role (primary/secondary) semantics can be relevant for any network topology, it describe the purpose of the network regardless of how the network topology is configured (e.g. IP addressing, enable layer2 / layer3 communication).

Reply:
> If that's on the roadmap then moving it up to spec may make sense

@trozet does primary network over localnet it planned to be supported in the future?

> Will there ever be more types of network? Would they always support primary/secondary semantics?

I think the interface role (primary/secondary) semantics can be relevant for any network topology, it describe the purpose of the network regardless of how the network topology is configured (e.g. IP addressing, enable layer2 / layer3 communication).

Reply:
I think it is possible to have more roles in the future, and we also know that role values may depend on the chosen topology type. But at the same time, network role semantically is not a topology configuration, but a pod attachment configuration, so it kinda makes sense outside the topology spec.
Considering this is just one field, we could bear extra validation based on topology type, but I am also wondering if `role` can be a part of discriminated union `UserDefinedNetworkSpec` as a non-discriminated member, like so
```
spec:
  role: Primary
  topology: Layer3
  layer3: <spec>
```
or do we need one extra layer like this
```
spec:
  role: Primary
  topology:
    type: Layer3
    layer3: <spec>
```

Reply:
> > If that's on the roadmap then moving it up to spec may make sense
> 
> @trozet does primary network over localnet it planned to be supported in the future?

I know I'm not Tim, but, from my perspective, that use case doesn't make sense. Localnet already has north/south traffic capabilities - depending on the configuration provided by the network admin.

Hence, it doesn't make sense to have routed egress for it. Likewise, I don't see a need to expose ports via services. Everything is automatically exposed. Should the user want to "hide" anything, they can use multi-net policies.

I can't speak for the other features though - egressIP / egress firewall / egress QoS / egress services. 

Maybe egress qos could be interesting ?... Still, I'd say that feature can be manually configured in the network by the admin.

> 
> > Will there ever be more types of network? Would they always support primary/secondary semantics?
> 
> I think the interface role (primary/secondary) semantics can be relevant for any network topology, it describe the purpose of the network regardless of how the network topology is configured (e.g. IP addressing, enable layer2 / layer3 communication).

Reply:
I made an attempt to extract `role` from the spec, and it is not pretty. There are multiple validation dependencies for topology spec based on role (sometimes together with the topology type!), e.g.
- JoinSubnets is only supported for Primary network
- Subnets is required for Primary Layer2 topology
- LocalNet topology is only supported for Secondary network

so it seems like network role is closely related to the topology and allowed configurations, and maybe it makes sense to just leave it where it is.
Thoughts?

Reply:
The mentioned scenarios should be validated in both cases, right?
Can we make use of the union discriminator field (the one that has the type name to use) to do the validations?
Please see this [example](https://playcel.undistro.io/?content=H4sIAAAAAAAAA6WSUUvDMBDHv8rfDIqCdkWfLPTJJ2GI4GteQnrrgsmlpNmkOL%2B7abtORBTHkpdc7u7%2Fu8vlXWiyohTLJR42pF87mDXihqA9R2WYAoxTDXVQgRBV01CNtbfWvxluxsiOnOJoNHYUOuM5lyz5smtJ58FbQlVBiueQZEIvBbIMoy%2F61lvf9Hns20PQymtlnyhKcYX9XvKXxsV3DXEtahXVI7fbmIpfpGqdI46SF9gpa2pMyaVkYBAoMSeni5k8epOZ%2BCVWqqdwO93Y8XxwAy5uS9wXRTE0dh7g7t%2BAH%2FovlEZS%2F0mYn29iDBZT%2FI1ieGrk5hiJbERV7dzJqR2ewB82jivN0%2Fma0iiH3%2FjxCQopSI6VAgAA) for  "LocalNet topology is only supported for Secondary network"  scenario, let me know what you think

Reply:
I looked at it again and tired to play with additional CEL rules, the fact there are multiple config types CEL rule should account for it makes validation way more complicated comparing to flat structure

Reply:
Taking into consideration:
- uncertainty about Localnet future as a primary network
- potentially new role types with topology dependencies
- existing dependencies between topology spec and network role
- uglier and more complicated CEL rules with `role` outside the topology spec

I will leave `role` as a part of topology spec. Thank you all for your comments, it was a great discussion!